File: python/visio_schema/wire/recordings.py
"""Pull recorded sessions off a device: list, open, receive, delete. Owns no bus connection.

Control rides the existing Command / CommandResult pair (``ListRecordings``,
``OpenRecordingFile``, ``DeleteRecording``). The file bytes do NOT ride the bus:
an ``OpenRecordingFile`` answer names a TCP port on the device's USB-NCM
address, and the host connects, reads until the device closes, and never
writes. Exactly ``length`` bytes means the range arrived whole; fewer means
open again at the bytes received. TCP carries order, delivery and flow control,
so nothing here re-implements them. The contract is
docs/protocol/recordings_pull.md.

Commands are transport-agnostic: every call takes ``run(command, timeout) ->
CommandResult``. The caller stamps ``command_id``, sends on COMMAND and returns
the matching result from ``/<device>/command_result``. `receive` opens its own
socket, because the byte path is a plain socket by design.
"""
from __future__ import annotations

import socket
from collections.abc import Callable
from typing import NamedTuple

from visio_schema.v1.control import command_pb2, command_result_pb2

__all__ = [
    "COMMAND_TIMEOUT_S",
    "DEFAULT_PORT",
    "MAX_NAME_BYTES",
    "MAX_OPENS_WITHOUT_PROGRESS",
    "PULL_QUIESCE_RULES",
    "STALL_TIMEOUT_S",
    "Received",
    "RecordingsError",
    "delete_recording",
    "delete_recording_command",
    "list_all_recordings",
    "list_recordings",
    "list_recordings_command",
    "list_session_files",
    "open_recording_file",
    "open_recording_file_command",
    "pull_file",
    "receive",
]

#: Where the device's file sender listens, on its USB-NCM address only. The
#: open reports the port; this is the value current firmware uses.
DEFAULT_PORT = 50002

#: How long to wait for one CommandResult.
COMMAND_TIMEOUT_S = 8.0

#: No byte for this long on the data socket means the transfer is dead; the
#: bytes that did arrive are kept and the next open resumes after them.
STALL_TIMEOUT_S = 10.0

#: Opens in a row that deliver nothing before a pull gives up.
MAX_OPENS_WITHOUT_PROGRESS = 5

#: What a host drops on its own bus link while it pulls (SetStreamPolicy rules,
#: section 7): the bulk streams — video, IMU samples, audio. Restore on every
#: exit. The C++ and TS twins spell the same tuple.
PULL_QUIESCE_RULES = ("**/camera/*", "**/imu/*/raw", "**/imu/*/quat", "**/audio/*")

#: Session names, file names and cursors are all capped at 63 UTF-8 bytes
#: (nanopb.options max_size:64). An oversized string does not truncate on the
#: device, it fails the whole decode and the device answers nothing, so the cap
#: is enforced here before anything is sent.
MAX_NAME_BYTES = 63

_RECV_BUFFER_BYTES = 1 << 20

Runner = Callable[[command_pb2.Command, float], command_result_pb2.CommandResult]
Sink = Callable[[int, bytes], None]


class RecordingsError(RuntimeError):
    """The device refused a command (``code`` is CommandResult.error_code), or a
    pull could not finish (``no_progress``, ``protocol``)."""

    def __init__(self, code: str, message: str = ""):
        super().__init__(f"{code}: {message}" if message else code)
        self.code = code


class Received(NamedTuple):
    """What one `receive` delivered. ``error`` is why it stopped short, if it did."""

    received: int
    error: OSError | None


def _check_name(what: str, name: str, *, required: bool = True) -> None:
    if not name:
        if required:
            raise ValueError(f"{what} is empty")
        return
    raw = name.encode()
    if len(raw) > MAX_NAME_BYTES:
        raise ValueError(f"{what} is {len(raw)} bytes; the limit is {MAX_NAME_BYTES}")
    # A leading dot also rejects "." and ".." — the C++ and TS twins say so too.
    if "/" in name or "\0" in name or name.startswith("."):
        raise ValueError(f"{what} {name!r} is not a plain name")


def list_recordings_command(*, limit: int = 0, cursor: str = "", session_name: str = "",
                            target_device: str = "") -> command_pb2.Command:
    """A ListRecordings. No cursor and no session = the original newest-``limit`` call."""
    if len(cursor.encode()) > MAX_NAME_BYTES:
        raise ValueError(f"cursor is {len(cursor.encode())} bytes; the limit is {MAX_NAME_BYTES}")
    _check_name("session_name", session_name, required=False)
    cmd = command_pb2.Command(target_device=target_device)
    body = cmd.list_recordings
    body.SetInParent()
    body.limit = limit
    body.cursor = cursor
    body.session_name = session_name
    return cmd


def open_recording_file_command(session_name: str, file_name: str, *, offset: int = 0,
                                expect_size: int = 0, expect_mtime_ns: int = 0,
                                target_device: str = "") -> command_pb2.Command:
    """An OpenRecordingFile. ``expect_*`` both 0 = unchecked."""
    _check_name("session_name", session_name)
    _check_name("file_name", file_name)
    cmd = command_pb2.Command(target_device=target_device)
    body = cmd.open_recording_file
    body.session_name = session_name
    body.file_name = file_name
    body.offset = offset
    body.expect_size = expect_size
    body.expect_mtime_ns = expect_mtime_ns
    return cmd


def delete_recording_command(session_name: str, *, target_device: str) -> command_pb2.Command:
    """A DeleteRecording. The device refuses a broadcast delete, so a target is required."""
    _check_name("session_name", session_name)
    if not target_device:
        raise ValueError("delete_recording needs target_device: a broadcast delete is refused")
    cmd = command_pb2.Command(target_device=target_device)
    cmd.delete_recording.session_name = session_name
    return cmd


def _run(run: Runner, cmd: command_pb2.Command, timeout: float) -> command_result_pb2.CommandResult:
    result = run(cmd, timeout)
    if not result.ok:
        raise RecordingsError(result.error_code or "failed", result.error_message)
    return result


def _payload(result: command_result_pb2.CommandResult, want: str, what: str):
    got = result.WhichOneof("payload")
    if got != want:
        raise RecordingsError("protocol", f"{what} answered payload {got!r}, expected {want!r}")
    return getattr(result, want)


def list_recordings(run: Runner, *, limit: int = 0, cursor: str = "", target_device: str = "",
                    timeout: float = COMMAND_TIMEOUT_S) -> command_result_pb2.RecordingsList:
    """One page of sessions, newest first, without files."""
    result = _run(run, list_recordings_command(limit=limit, cursor=cursor,
                                               target_device=target_device), timeout)
    return _payload(result, "recordings", "ListRecordings")


def list_all_recordings(
    run: Runner, *, page_limit: int = 0, target_device: str = "",
    timeout: float = COMMAND_TIMEOUT_S,
) -> list[command_result_pb2.RecordingEntry]:
    """Every session, following cursors to the last page."""
    entries: list[command_result_pb2.RecordingEntry] = []
    cursor = ""
    seen: set[str] = set()
    while True:
        page = list_recordings(run, limit=page_limit, cursor=cursor,
                               target_device=target_device, timeout=timeout)
        entries.extend(page.recordings)
        if not page.next_cursor:
            return entries
        if page.next_cursor in seen:
            raise RecordingsError("protocol", f"cursor {page.next_cursor!r} did not advance")
        seen.add(page.next_cursor)
        cursor = page.next_cursor


def list_session_files(run: Runner, session_name: str, *, target_device: str = "",
                       timeout: float = COMMAND_TIMEOUT_S) -> command_result_pb2.RecordingEntry:
    """One session with its files (``RecordingEntry.files``)."""
    result = _run(run, list_recordings_command(session_name=session_name,
                                               target_device=target_device), timeout)
    page = _payload(result, "recordings", "ListRecordings")
    if len(page.recordings) != 1 or page.recordings[0].name != session_name:
        raise RecordingsError("protocol", f"asked for {session_name!r}, got "
                              f"{[e.name for e in page.recordings]!r}")
    return page.recordings[0]


def open_recording_file(run: Runner, session_name: str, file_name: str, *, offset: int = 0,
                        expect_size: int = 0, expect_mtime_ns: int = 0, target_device: str = "",
                        timeout: float = COMMAND_TIMEOUT_S) -> command_result_pb2.RecordingFileOpen:
    """Open one file for reading from ``offset``; the answer says where and how much to read."""
    cmd = open_recording_file_command(session_name, file_name, offset=offset,
                                      expect_size=expect_size, expect_mtime_ns=expect_mtime_ns,
                                      target_device=target_device)
    opened = _payload(_run(run, cmd, timeout), "file_open", "OpenRecordingFile")
    if opened.offset != offset or opened.offset + opened.length != opened.file_size:
        raise RecordingsError("protocol", f"open for offset {offset} answered offset "
                              f"{opened.offset}, length {opened.length}, size {opened.file_size}")
    return opened


def receive(address: str, opened: command_result_pb2.RecordingFileOpen, sink: Sink, *,
            stall_timeout: float = STALL_TIMEOUT_S) -> Received:
    """Read one opened range from the device into ``sink(offset, data)``.

    Connects, reads until the device closes or ``opened.length`` bytes have
    arrived, and never writes. A short read is not an error here — a refused
    connection, a reset, a stall and an early close all return what arrived,
    with the reason in ``error``; the caller opens again at the bytes received.
    Exceptions from ``sink`` propagate: a full disk is not a reason to retry.

    This reads only the data socket. The caller keeps its bus link read for the
    duration (section 4): the device aborts a bus connection left unread 10 s.
    """
    try:
        sock = socket.create_connection((address, opened.port), timeout=stall_timeout)
    except OSError as err:
        return Received(0, err)
    received = 0
    buf = bytearray(_RECV_BUFFER_BYTES)
    view = memoryview(buf)
    with sock:
        sock.settimeout(stall_timeout)
        while received < opened.length:
            # Only the socket is guarded: an error from the sink is the caller's
            # (a full disk must stop the pull, not look like a dropped link).
            try:
                n = sock.recv_into(view, min(len(buf), opened.length - received))
            except OSError as err:  # includes socket.timeout
                return Received(received, err)
            if n == 0:
                break
            sink(opened.offset + received, bytes(view[:n]))
            received += n
    return Received(received, None)


def pull_file(run: Runner, address: str, session_name: str, file_name: str, sink: Sink, *,
              offset: int = 0, expect_size: int = 0, expect_mtime_ns: int = 0,
              target_device: str = "", on_progress: Callable[[int, int], None] | None = None,
              stall_timeout: float = STALL_TIMEOUT_S, timeout: float = COMMAND_TIMEOUT_S,
              max_opens_without_progress: int = MAX_OPENS_WITHOUT_PROGRESS,
              ) -> command_result_pb2.RecordingFileOpen:
    """Pull ``file_name`` from ``offset`` to its end into ``sink(offset, data)``.

    Opens, receives, and on a short read opens again at the bytes received,
    pinning the file's size and mtime from the first open so a file that
    changes between attempts fails with "changed" instead of being spliced.
    Returns the last open (its ``file_size`` and ``mtime_ns`` identify the
    file). Device refusals raise RecordingsError with the device's code.
    """
    idle_opens = 0
    last_error: OSError | None = None
    while True:
        opened = open_recording_file(run, session_name, file_name, offset=offset,
                                     expect_size=expect_size, expect_mtime_ns=expect_mtime_ns,
                                     target_device=target_device, timeout=timeout)
        expect_size, expect_mtime_ns = opened.file_size, opened.mtime_ns
        if opened.length == 0:
            return opened
        got = receive(address, opened, sink, stall_timeout=stall_timeout)
        offset += got.received
        if on_progress is not None:
            on_progress(offset, opened.file_size)
        if offset == opened.file_size:
            return opened
        if got.received:
            idle_opens = 0
        else:
            idle_opens += 1
            last_error = got.error
            if idle_opens >= max_opens_without_progress:
                raise RecordingsError("no_progress", f"{idle_opens} opens delivered no bytes "
                                      f"at offset {offset} (last: {last_error!r})")


def delete_recording(run: Runner, session_name: str, *, target_device: str,
                     timeout: float = COMMAND_TIMEOUT_S) -> None:
    """Delete one whole session. Refusals raise RecordingsError with the device's code."""
    _run(run, delete_recording_command(session_name, target_device=target_device), timeout)

File: ts/src/wire/recordings.ts
/**
 * Pull recorded sessions off a device: list, open, receive, delete. Owns no bus
 * connection.
 *
 * The TypeScript twin of `visio_schema/wire/recordings.py`. Control rides the
 * existing Command / CommandResult pair (ListRecordings, OpenRecordingFile,
 * DeleteRecording). The file bytes do NOT ride the bus: an open names a TCP
 * port, and the host connects to it on the device address it sent the open to,
 * reads until the device closes, and never writes. Any direct IP link serves a
 * pull (USB-NCM or Wi-Fi). Exactly `length` bytes means the range arrived
 * whole; fewer means open again at the bytes received. The contract is
 * docs/protocol/recordings_pull.md.
 *
 * Both seams are injected. `run` sends one Command and resolves with its
 * CommandResult (the caller stamps commandId and matches the reply). `receive`
 * reads one opened range into wherever the bytes belong — a React Native app
 * does that natively, straight to a file, and Node does it with `node:net` —
 * so this module imports neither.
 */
import { create, fromBinary, toBinary } from '@bufbuild/protobuf';

import {
  CommandSchema,
  DeleteRecordingSchema,
  ListRecordingsSchema,
  OpenRecordingFileSchema,
  type Command,
} from '../gen/visio_schema/v1/control/command_pb.js';
import {
  CommandResultSchema,
  type CommandResult,
  type RecordingEntry,
  type RecordingFileOpen,
  type RecordingsList,
} from '../gen/visio_schema/v1/control/command_result_pb.js';

/** Where current firmware's file sender listens (the open reports the port). */
export const DEFAULT_PORT = 50002;
export const COMMAND_TIMEOUT_S = 8.0;
/** No byte for this long on the data socket: the transfer is dead; resume. */
export const STALL_TIMEOUT_S = 10.0;
export const MAX_OPENS_WITHOUT_PROGRESS = 5;
/**
 * What a host drops on its own bus link while it pulls (SetStreamPolicy rules,
 * recordings_pull.md §7): video, IMU samples, audio. Restore on every exit.
 */
export const PULL_QUIESCE_RULES = ['**/camera/*', '**/imu/*/raw', '**/imu/*/quat', '**/audio/*'] as const;
/**
 * Names and the cursor are capped at 63 UTF-8 bytes (nanopb max_size:64). An
 * oversized string fails the device's whole decode and it answers nothing, so
 * the cap is enforced before anything is sent.
 */
export const MAX_NAME_BYTES = 63;

/** The device refused (`code` = CommandResult.errorCode), or a pull could not finish. */
export class RecordingsError extends Error {
  readonly code: string;
  constructor(code: string, message = '') {
    super(message ? `${code}: ${message}` : code);
    this.name = 'RecordingsError';
    this.code = code;
  }
}

/** Send one Command (stamp commandId first) and resolve with its CommandResult. */
export type RunCommand = (command: Command, timeoutS: number) => Promise<CommandResult>;

export interface Received {
  /** Bytes of the range that reached their destination. */
  readonly received: bigint;
  /** Why it stopped short, if it did — for the error message only. */
  readonly error?: string;
}

/**
 * Read one opened range: connect to `open.port` on the device address the open
 * was sent to, store bytes from `open.offset` in order, stop at `open.length`
 * or when the device closes. A short read is a value, not a throw. Throw only
 * for a failure retrying cannot fix (a full disk), and that ends the pull.
 * The bus link must keep being read meanwhile (recordings_pull.md §4).
 */
export type Receive = (open: RecordingFileOpen, stallTimeoutS: number) => Promise<Received>;

const utf8Bytes = (s: string): number => new TextEncoder().encode(s).length;

function checkName(what: string, name: string, required = true): void {
  if (!name) {
    if (required) throw new RangeError(`${what} is empty`);
    return;
  }
  const n = utf8Bytes(name);
  if (n > MAX_NAME_BYTES) throw new RangeError(`${what} is ${n} bytes; the limit is ${MAX_NAME_BYTES}`);
  if (name.includes('/') || name.includes('\0') || name.startsWith('.')) {
    throw new RangeError(`${what} ${JSON.stringify(name)} is not a plain name`);
  }
}

export interface Addressed {
  targetDevice?: string;
}

/** A ListRecordings. No cursor and no session = the original newest-`limit` call. */
export function listRecordingsCommand(
  o: Addressed & { limit?: number; cursor?: string; sessionName?: string } = {},
): Command {
  const cursor = o.cursor ?? '';
  if (utf8Bytes(cursor) > MAX_NAME_BYTES) throw new RangeError(`cursor is over ${MAX_NAME_BYTES} bytes`);
  checkName('sessionName', o.sessionName ?? '', false);
  return create(CommandSchema, {
    targetDevice: o.targetDevice ?? '',
    body: {
      case: 'listRecordings',
      value: create(ListRecordingsSchema, {
        limit: o.limit ?? 0,
        cursor,
        sessionName: o.sessionName ?? '',
      }),
    },
  });
}

export interface OpenOptions extends Addressed {
  offset?: bigint;
  /** Both 0n = unchecked. */
  expectSize?: bigint;
  expectMtimeNs?: bigint;
}

export function openRecordingFileCommand(sessionName: string, fileName: string, o: OpenOptions = {}): Command {
  checkName('sessionName', sessionName);
  checkName('fileName', fileName);
  return create(CommandSchema, {
    targetDevice: o.targetDevice ?? '',
    body: {
      case: 'openRecordingFile',
      value: create(OpenRecordingFileSchema, {
        sessionName,
        fileName,
        offset: o.offset ?? 0n,
        expectSize: o.expectSize ?? 0n,
        expectMtimeNs: o.expectMtimeNs ?? 0n,
      }),
    },
  });
}

/** A DeleteRecording. The device refuses a broadcast delete, so a target is required. */
export function deleteRecordingCommand(sessionName: string, targetDevice: string): Command {
  checkName('sessionName', sessionName);
  if (!targetDevice) throw new RangeError('deleteRecording needs targetDevice: a broadcast delete is refused');
  return create(CommandSchema, {
    targetDevice,
    body: { case: 'deleteRecording', value: create(DeleteRecordingSchema, { sessionName }) },
  });
}

export const encodeCommand = (c: Command): Uint8Array => toBinary(CommandSchema, c);
export const decodeResult = (bytes: Uint8Array): CommandResult => fromBinary(CommandResultSchema, bytes);

async function runOk(run: RunCommand, cmd: Command, timeoutS: number): Promise<CommandResult> {
  const result = await run(cmd, timeoutS);
  if (!result.ok) throw new RecordingsError(result.errorCode || 'failed', result.errorMessage);
  return result;
}

function listing(result: CommandResult): RecordingsList {
  if (result.payload.case !== 'recordings') {
    throw new RecordingsError('protocol', `ListRecordings answered payload ${String(result.payload.case)}`);
  }
  return result.payload.value;
}

export async function listRecordings(
  run: RunCommand,
  o: Addressed & { limit?: number; cursor?: string; timeoutS?: number } = {},
): Promise<RecordingsList> {
  const cmd = listRecordingsCommand({ limit: o.limit, cursor: o.cursor, targetDevice: o.targetDevice });
  return listing(await runOk(run, cmd, o.timeoutS ?? COMMAND_TIMEOUT_S));
}

/** Every session, following cursors to the last page. */
export async function listAllRecordings(
  run: RunCommand,
  o: Addressed & { pageLimit?: number; timeoutS?: number } = {},
): Promise<RecordingEntry[]> {
  const out: RecordingEntry[] = [];
  const seen = new Set<string>();
  let cursor = '';
  for (;;) {
    const page = await listRecordings(run, { limit: o.pageLimit, cursor, targetDevice: o.targetDevice, timeoutS: o.timeoutS });
    out.push(...page.recordings);
    if (!page.nextCursor) return out;
    if (seen.has(page.nextCursor)) throw new RecordingsError('protocol', `cursor ${page.nextCursor} did not advance`);
    seen.add(page.nextCursor);
    cursor = page.nextCursor;
  }
}

/** One session with its files (`RecordingEntry.files`). */
export async function listSessionFiles(
  run: RunCommand,
  sessionName: string,
  o: Addressed & { timeoutS?: number } = {},
): Promise<RecordingEntry> {
  const cmd = listRecordingsCommand({ sessionName, targetDevice: o.targetDevice });
  const page = listing(await runOk(run, cmd, o.timeoutS ?? COMMAND_TIMEOUT_S));
  const only = page.recordings.length === 1 ? page.recordings[0] : undefined;
  if (!only || only.name !== sessionName) {
    throw new RecordingsError('protocol', `asked for ${sessionName}, got ${page.recordings.map((e) => e.name).join(',')}`);
  }
  return only;
}

/** Open one file from `offset`; the answer says where and how much to read. */
export async function openRecordingFile(
  run: RunCommand,
  sessionName: string,
  fileName: string,
  o: OpenOptions & { timeoutS?: number } = {},
): Promise<RecordingFileOpen> {
  const offset = o.offset ?? 0n;
  const result = await runOk(run, openRecordingFileCommand(sessionName, fileName, o), o.timeoutS ?? COMMAND_TIMEOUT_S);
  if (result.payload.case !== 'fileOpen') {
    throw new RecordingsError('protocol', `OpenRecordingFile answered payload ${String(result.payload.case)}`);
  }
  const opened = result.payload.value;
  if (opened.offset !== offset || opened.offset + opened.length !== opened.fileSize) {
    throw new RecordingsError(
      'protocol',
      `open for offset ${offset} answered offset ${opened.offset}, length ${opened.length}, size ${opened.fileSize}`,
    );
  }
  return opened;
}

export interface PullIo {
  run: RunCommand;
  receive: Receive;
  onProgress?: (offset: bigint, size: bigint) => void;
}

export interface PullOptions extends OpenOptions {
  stallTimeoutS?: number;
  commandTimeoutS?: number;
  maxOpensWithoutProgress?: number;
}

/**
 * Pull `fileName` from `offset` to its end. On a short read, open again at the
 * bytes received with the size and mtime of the first open pinned, so a file
 * that changes between attempts fails "changed" instead of being spliced.
 * Resolves with the last open (its fileSize and mtimeNs identify the file).
 */
export async function pullFile(
  io: PullIo,
  sessionName: string,
  fileName: string,
  o: PullOptions = {},
): Promise<RecordingFileOpen> {
  let offset = o.offset ?? 0n;
  let expectSize = o.expectSize ?? 0n;
  let expectMtimeNs = o.expectMtimeNs ?? 0n;
  const maxIdle = o.maxOpensWithoutProgress ?? MAX_OPENS_WITHOUT_PROGRESS;
  let idleOpens = 0;
  for (;;) {
    const opened = await openRecordingFile(io.run, sessionName, fileName, {
      offset,
      expectSize,
      expectMtimeNs,
      targetDevice: o.targetDevice,
      timeoutS: o.commandTimeoutS,
    });
    expectSize = opened.fileSize;
    expectMtimeNs = opened.mtimeNs;
    if (opened.length === 0n) return opened;
    const got = await io.receive(opened, o.stallTimeoutS ?? STALL_TIMEOUT_S);
    offset += got.received;
    io.onProgress?.(offset, opened.fileSize);
    if (offset === opened.fileSize) return opened;
    if (got.received > 0n) {
      idleOpens = 0;
    } else {
      idleOpens += 1;
      if (idleOpens >= maxIdle) {
        throw new RecordingsError('no_progress', `${idleOpens} opens delivered no bytes at offset ${offset} (last: ${got.error ?? 'closed'})`);
      }
    }
  }
}

/** Delete one whole session. Refusals throw RecordingsError with the device's code. */
export async function deleteRecording(
  run: RunCommand,
  sessionName: string,
  targetDevice: string,
  o: { timeoutS?: number } = {},
): Promise<void> {
  await runOk(run, deleteRecordingCommand(sessionName, targetDevice), o.timeoutS ?? COMMAND_TIMEOUT_S);
}
